edfio/edf.py

```diff
@@ -1019,37 +1023,99 @@ def update_data_record_duration(
             - "truncate": Truncate the data to the previous compatible duration (might
               lead to loss of data)
         """
+        if data_record_duration == self.data_record_duration:
+            return
+        if not any(
+            signal for signal in self.signals if signal not in self._annotation_signals
+        ):
+            raise ValueError(
+                "Data record duration must be zero for annotation-only files"
+            )
         for signal in self.signals:
+            if signal in self._annotation_signals:
+                continue
             spr = signal.sampling_frequency * data_record_duration
+            if spr <= 0:
+                raise ValueError(
+                    f"Data record duration must be positive, got {data_record_duration}"
+                )
             if spr % 1:
                 raise ValueError(
                     f"Cannot set data record duration to {data_record_duration}: Incompatible sampling frequency {signal.sampling_frequency} Hz"
                 )
+
+        num_data_records = self._pad_or_truncate_signals(data_record_duration, method)
+        self._update_record_duration_in_annotation_signals(
+            data_record_duration, num_data_records
+        )
+        self._data_record_duration = data_record_duration
+        self._num_data_records = Edf.num_data_records.encode(num_data_records)
+
+    def _pad_or_truncate_signals(
+        self, data_record_duration: float, method: Literal["strict", "pad", "truncate"]
+    ) -> int:
         if method == "pad":
             new_duration = (
                 ceil(self.duration / data_record_duration) * data_record_duration
             )
             self._pad_or_truncate_data(new_duration)
-        elif method == "truncate":
+            return round(new_duration / data_record_duration)
+        if method == "truncate":
             new_duration = (
                 floor(self.duration / data_record_duration) * data_record_duration
             )
             self._pad_or_truncate_data(new_duration)
-        self._data_record_duration = data_record_duration
-        self._set_num_data_records_with_signals(self.signals)
+            return round(new_duration / data_record_duration)
+        return _calculate_num_data_records(self.duration, data_record_duration)
+
+    def _update_record_duration_in_annotation_signals(
+        self, data_record_duration: float, num_data_records: int
+    ) -> None:
+        first_annotation_signal = True
+        for idx, signal in enumerate(self._signals):
+            if signal not in self._annotation_signals:
+                continue
+            annotations = []
+            for data_record in signal._digital.reshape(
+                (-1, signal.samples_per_data_record)
+            ):
+                annot_dr = _EdfAnnotationsDataRecord.from_bytes(data_record.tobytes())
+                if first_annotation_signal:
```

Comment:
The flag `first_annotation_signal` could be avoided using this:
```suggestion
                if signal is self._timekeeping_signal:
```